File: web/src/lib/workspace.ts
export type WorkspaceState = 'unconfigured' | 'ready' | 'invalid'

export interface AgentProfile {
  id: string
  label: string
  requires_primary_directory: boolean
  supports_additional_directories: boolean
}

export interface WorkspaceScope {
  id: string
  label: string
  primary_directory: string
}

export interface WorkspaceStatus {
  state: WorkspaceState
  profile: AgentProfile
  workspace: WorkspaceScope | null
}

export interface AgentSettingsStatus {
  profiles: AgentProfile[]
  selected_profile: AgentProfile
}

export interface AgentSelectionResult {
  settings: AgentSettingsStatus
  runtime: LocalRuntimeStatus
}

export type BrowseOperationState = 'picking' | 'ready' | 'cancelled' | 'failed'

export interface BrowseOperationStatus {
  operation_id: string
  state: BrowseOperationState
  workspace?: WorkspaceStatus | null
  error?: string | null
}

export type BrowseWorkspaceOutcome = { state: 'ready'; workspace: WorkspaceStatus } | { state: 'cancelled' }

export type LocalRuntimeState = 'configuration_required' | 'starting' | 'authentication_required' | 'ready' | 'failed'

export interface LocalRuntimeStatus {
  state: LocalRuntimeState
  workspace: WorkspaceStatus
  error: string | null
}

export type ClaudeAuthState = 'signed_out' | 'waiting' | 'signed_in' | 'failed'

export interface ClaudeAuthStatus {
  state: ClaudeAuthState
  error: string | null
}

export const CODEX_PROFILE: AgentProfile = {
  id: 'codex',
  label: 'Codex',
  requires_primary_directory: true,
  supports_additional_directories: false,
}

export function workspaceNeedsConfiguration(status: WorkspaceStatus): boolean {
  return status.state !== 'ready' || status.workspace === null
}

File: server/src/acp_runtime/workspace.py
"""Durable selection of one ACP Workspace Scope."""

import hashlib
import json
import os
import tempfile
from dataclasses import asdict, dataclass
from pathlib import Path
from typing import Callable, Literal


@dataclass(frozen=True)
class AgentProfile:
    """Capabilities that shape Workspace Settings without naming a UI vendor."""

    id: str = "codex"
    label: str = "Codex"
    requires_primary_directory: bool = True
    supports_additional_directories: bool = False


CODEX_PROFILE = AgentProfile()
_PROJECT_FOLDER_ERROR = "Project Folder must be an absolute existing directory"


def resolve_project_folder(path: str) -> str:
    """Return one resolved absolute directory for Workspace and ACP boundaries."""
    candidate = Path(path).expanduser()
    if not candidate.is_absolute():
        raise ValueError(_PROJECT_FOLDER_ERROR)
    try:
        resolved = candidate.resolve(strict=True)
    except (OSError, RuntimeError) as exc:
        raise ValueError(_PROJECT_FOLDER_ERROR) from exc
    if not resolved.is_dir():
        raise ValueError(_PROJECT_FOLDER_ERROR)
    return str(resolved)


@dataclass(frozen=True)
class WorkspaceScope:
    """ACP session context for one resolved Project Folder."""

    id: str
    label: str
    primary_directory: str


@dataclass(frozen=True)
class WorkspaceStatus:
    state: Literal["unconfigured", "ready", "invalid"]
    profile: AgentProfile
    workspace: WorkspaceScope | None


class WorkspaceConfigStore:
    """Persist exactly one Workspace Scope as an atomic local JSON record."""

    SCHEMA_VERSION = "1.0"

    def __init__(self, path: Path) -> None:
        self.path = path

    @classmethod
    def default(cls) -> "WorkspaceConfigStore":
        state_directory = os.getenv("VOICE_ACP_STATE_DIR")
        root = (
            Path(state_directory).expanduser()
            if state_directory
            else Path.home()
            / "Library"
            / "Application Support"
            / "Agora Voice ACP"
        )
        return cls(root / "workspace.json")

    def load(self) -> WorkspaceScope | None:
        if not self.path.exists():
            return None
        payload = json.loads(self.path.read_text(encoding="utf-8"))
        if payload.get("schema_version") != self.SCHEMA_VERSION:
            raise ValueError("Unsupported Workspace configuration version")
        return WorkspaceScope(
            id=str(payload["id"]),
            label=str(payload["label"]),
            primary_directory=str(payload["primary_directory"]),
        )

    def save(self, workspace: WorkspaceScope) -> None:
        self.path.parent.mkdir(parents=True, exist_ok=True)
        payload = {
            "schema_version": self.SCHEMA_VERSION,
            **asdict(workspace),
        }
        descriptor, temporary_name = tempfile.mkstemp(
            dir=self.path.parent,
            prefix=f".{self.path.name}.",
            suffix=".tmp",
        )
        temporary_path = Path(temporary_name)
        try:
            with os.fdopen(descriptor, "w", encoding="utf-8") as stream:
                json.dump(payload, stream, ensure_ascii=False, sort_keys=True)
                stream.write("\n")
                stream.flush()
                os.fsync(stream.fileno())
            os.replace(temporary_path, self.path)
        except BaseException:
            temporary_path.unlink(missing_ok=True)
            raise

    def clear(self) -> None:
        self.path.unlink(missing_ok=True)


class WorkspaceService:
    """Validate and report the selected Workspace Scope."""

    def __init__(
        self,
        store: WorkspaceConfigStore,
        profile_provider: Callable[[], AgentProfile] | None = None,
    ) -> None:
        self.store = store
        self._profile_provider = profile_provider or (lambda: CODEX_PROFILE)

    def status(self) -> WorkspaceStatus:
        workspace = self.store.load()
        if workspace is None:
            return WorkspaceStatus(
                state="unconfigured",
                profile=self._profile_provider(),
                workspace=None,
            )
        path = Path(workspace.primary_directory)
        state = "ready" if path.is_absolute() and path.is_dir() else "invalid"
        return WorkspaceStatus(
            state=state,
            profile=self._profile_provider(),
            workspace=workspace,
        )

    def select(self, path: str) -> WorkspaceStatus:
        resolved = resolve_project_folder(path)

        workspace = WorkspaceScope(
            id=hashlib.sha256(os.fsencode(resolved)).hexdigest()[:24],
            label=Path(resolved).name or resolved,
            primary_directory=resolved,
        )
        self.store.save(workspace)
        return WorkspaceStatus(
            state="ready",
            profile=self._profile_provider(),
            workspace=workspace,
        )

    def restore(self, previous: WorkspaceStatus) -> WorkspaceStatus:
        """Restore one status snapshot after a replacement activation fails."""
        if previous.workspace is None:
            self.store.clear()
        else:
            self.store.save(previous.workspace)
        return self.status()

    def clear(self) -> WorkspaceStatus:
        self.store.clear()
        return WorkspaceStatus(
            state="unconfigured",
            profile=self._profile_provider(),
            workspace=None,
        )
